### DISCUSSION TITLE
Maiko: CHECK_INTERRUPT

### DISCUSSION THREAD

**Discussion by waywardmonkeys** [Interlisp/medley#166] [Q&A] [answered]
I have a question about `CHECK_INTERRUPT`.

It is defined as  (slightly reformatted for reading):

```
#define CHECK_INTERRUPT {\
    if ((UNSIGNED)CSTKPTR > Irq_Stk_Check) \
        goto check_interrupt; \
    } 
```

And now, let's take a look at an example usage of it ...

```
#define JUMPMACRO(x)    {\
    CHECK_INTERRUPT; \
    PCMACL += x; \
    nextop0; \
}
```

Now, if the interrupt actually fires and it does the `goto check_interrupt`, how does it get the `PCMACL` updated correctly? At the end of the `check_interrupt` block, it just does a `nextop0`. This seems to be an issue with more than just that macro, although most are related to jumps.

**Comment by nbriggs** [comment, marked as answer]
I think what happens (check me on this!) is that if it is actually executing Lisp code in the interrupt, it will save and restore the stack, but when it returns it will return to the same opcode that it was executing (from the JUMPMACRO) because the PC has not been updated, this time the interrupt check won't fire, and it will update the PC appropriately and end up in the real next opcode.

**Reply by waywardmonkeys** [reply]
I think that sounds correct. But now I want to audit things a bit more closely and perhaps add some comments to the code.

Thanks!
